packages/oauth-js/src/client.ts

```diff
@@ -0,0 +1,361 @@
+
+    const url = buildAuthorizeUrl(this.config.domain, {
+      clientId: this.config.clientId,
+      redirectUri: this.config.callbackUri || "",
+      state: stateCode,
+      challenge: challenge,
+      challengeMethod: "S256",
+      responseType: "code_challenge",
+      scope: this.config.scope,
+    });
+
+    window.location.assign(url);
+  }
+
+  async logout() {
+    const response = await this.post("logout", {
+      token: getRefreshToken(this.options),
+      token_type_hint: "refresh_token",
+    });
+
+    if (response.ok) {
+      this.authState = "INIT";
+    } else {
+      this.authState = "ERROR";
+    }
+    removeTokenCookies(this.options);
+    this.storage.removeItem(STATE_DATA_KEY);
+  }
+
+  async refresh() {
+    const initialState = this.authState;
+    const lock = new Lock();
+
+    try {
+      this.authState = "UPDATING";
+
+      if (await lock.acquireLock(REFRESH_LOCK_KEY)) {
+        const refreshToken = getRefreshToken(this.options);
+        const response = await this.post("token", {
+          client_id: this.config.clientId,
+          refresh_token: refreshToken,
+          grant_type: "refresh_token",
+          redirect_uri: this.config.callbackUri,
+          scope: this.config.scope || "",
+        });
+
+        this.processResponse(response);
+      } else {
+        console.warn("Could not aquire refresh lock.");
+      }
+    } catch (error: any) {
+      console.log(error);
+      if (initialState !== "INIT") {
+        // refresh expected to fail in initial state
+      } else {
+        this.authState = "ERROR";
+      }
+    } finally {
+      await lock.releaseLock(REFRESH_LOCK_KEY);
+    }
+  }
+
+  async introspect() {
+    const token = getSessionToken(this.options);
+    const response = await this.post("token/introspect", {
+      client_id: this.config.clientId,
+      token_type_hint: "access_token",
+      token,
+    });
+
+    return await response.json();
+  }
+
+  async userinfo() {
+    const response = await this.get("userinfo");
+
+    return await response.json();
+  }
+
+  async getToken() {
+    if (this.shouldTokenRefresh()) {
+      try {
+        await this.refresh();
+      } catch (err) {}
+    }
+
+    const accessToken = getSessionToken(this.options);
+
+    if (!accessToken) {
+      // throw error
+      this.authState = "ERROR";
+    }
+
+    return accessToken;
+  }
+
+  private checkTokenLife() {
+    const tokenExpire = getTokenExpire(this.options);
+
+    if (tokenExpire && isTokenExpiring(tokenExpire)) {
+      this.refresh();
+    }
+  }
+
+  private startTokenWatch() {
+    const intervalTime = REFRESH_CHECK_INTERVAL * 1000;
+
+    this.stopTokenWatch();
+
+    this.timer = window.setInterval(() => {
+      this.checkTokenLife();
+    }, intervalTime);
+  }
+
+  private stopTokenWatch() {
+    if (this.timer) {
+      clearInterval(this.timer);
+      this.timer = null;
+    }
+  }
+
+  private processResponse(data: any) {
```

Comment:
nit: maybe rename to be more specific on what type of response this is meant to be processing

Reply:
updated function name